csrc/preseg_passes/move_gather.cpp

```diff
@@ -0,0 +1,238 @@
+  // Check if the expression is a UnaryOp and a pointwise operation
+  return expr->isA<UnaryOp>() && ir_utils::isPointwiseTvOp(expr);
+}
+
+// For now we allow unary pointwise ops like cast/neg or a squeeze.
+// We only consider squeeze which has a single dim squeezed.
+std::optional<Expr*> getAllowedLookupTvDef(const GatherOp* gather_op) {
+  if (gather_op->lookupTv()->isFusionInput()) {
+    return std::nullopt;
+  }
+
+  auto producing_expr = gather_op->lookupTv()->definition();
+  NVF_ERROR(
+      producing_expr != nullptr, "Def for Lookup tensor view is nullptr.");
+
+  if (isUnaryPointwiseOp(producing_expr)) {
+    return producing_expr;
+  }
+
+  // We handle squeeze op where only one dim is squeezed.
+  if (producing_expr->isA<SqueezeOp>()) {
+    auto squeeze_op = producing_expr->as<SqueezeOp>();
+    if (std::count(
+            squeeze_op->getSqueezeDimFlags().begin(),
+            squeeze_op->getSqueezeDimFlags().end(),
+            true) == 1) {
+      return producing_expr;
+    }
+  }
+
+  return std::nullopt;
+}
+
+auto positionOfSqueezedDim(const std::vector<bool>& squeeze_dim_flags) {
+  auto it = std::find(squeeze_dim_flags.begin(), squeeze_dim_flags.end(), true);
+  return it != squeeze_dim_flags.end()
+      ? std::distance(squeeze_dim_flags.begin(), it)
+      : -1;
+}
+
+// If the def of lookUpTv is squeeze, and we're moving the squeeze to
+// after the gather/take_along_axis, then we need to add an unsqueeze
+// to the IndexTv
+auto prepareIndexTv(Fusion* fusion, GatherOp* gather_op, Expr* def) {
+  if (def->isA<SqueezeOp>()) {
+    auto squeeze_dim_flags = def->as<SqueezeOp>()->getSqueezeDimFlags();
+    auto dim = positionOfSqueezedDim(squeeze_dim_flags);
+    return unsqueeze(gather_op->indexTv(), dim);
+  }
+  return gather_op->indexTv();
+}
+
+auto addPostGatherSqueeze(
+    Fusion* fusion,
+    GatherOp* old_gather,
+    GatherOp* new_gather,
+    Expr* def) {
+  auto new_squeeze_output = squeeze(
+      new_gather->output(0)->as<TensorView>(),
+      def->as<SqueezeOp>()->getSqueezeDimFlags());
+
+  TransformReplay::selfReplay(
+      old_gather->output(0)->as<TensorView>()->domain(),
+      new_squeeze_output->domain());
+
+  return new_squeeze_output;
+}
+
+auto addPostGatherOp(Fusion* fusion, GatherOp* new_gather, Expr* def) {
+  IrCloner ir_cloner(fusion);
+
+  auto cloned_def = static_cast<Expr*>(def->clone(&ir_cloner));
+
+  // Replace the input with the output of the new gather.
+  cloned_def = ir_utils::replaceValInExprInputs(
+      cloned_def, cloned_def->input(0), new_gather->output(0));
+
+  // Since it's a unary pointwise op, the output should be like the input
+  // but type may vary as in the case of cast.
+  auto cloned_output_tv = ops::newValLike(
+      cloned_def->input(0)->as<TensorView>(), cloned_def->output(0)->dtype());
+  cloned_def =
+      ir_utils::transferDefinitionToNewOutputs(cloned_def, {cloned_output_tv});
+
+  TransformReplay::selfReplay(
```

Comment:
@jjsjann123 Does this selfReplay look reasonable? May be the output of the cloned_def (ex. cast) should resemble the output of the original gather.